wordpress-mobile/WordPress-Android
# Recommend the app

Fixes #15269 

https://user-images.githubusercontent.com/47797566/131883329-eb8aaaf7-a08a-450d-9f02-eba86671cf78.mov

The output as per the above has been already checked with @mattmiklic but cc-ing you fyi 🙇 . Note also that final implementation will add a share button to the about screen as well, but there has been already a discussion on slack (p1630508255058500?thread_ts=1630499387.055200&cid=C01CW1VMLAF-slack-hummingbird) and given the about screen is going to be reviewed soon we will insert the sharing option there.

This PR introduce (behind a feature flag for now) a `recommend the app` feature. When the feature flag is active, you get a `Share WordPress with a friend`menu item in the `Me Screen`. Tapping on the menu item an Android Share Sheet is opened that allows to copy the link or share with the apps that can intercept the ACTION_SEND in your device.

**Note 1:** the text and share link template is recovered from a dedicated API endpoint. The subject (where applicable) is currently hard coded as a string resource. Endpoint logic mimic iOS already deployed implementation, for this simple implementation we didn't create a store but implemented a Api calls provider. We can revisit this in a v2 in case we decide to add more fields and cache the template.

**Note 2:** the template is recovered only when the VM has not recovered it yet; in the video above for example you can see the first time the shimmer effect on the button is shown (fetching the API) but the second time being the VM still there the shimmer is not shown (using the already fetched template).

**Note 3:** the endpoint is not authenticated so everything should work also for pure self-hosted sites.

Below a few examples of the sharing operation result:

| Message | ![image](https://user-images.githubusercontent.com/47797566/131881133-9984aed4-823f-4c93-ba0a-053f5f21d796.png) |
|---|---|
| Mail | ![image](https://user-images.githubusercontent.com/47797566/131880898-ea792bd6-7483-4967-9635-e1b29f332684.png) |
| Twitter | ![image](https://user-images.githubusercontent.com/47797566/131885900-e40d5876-698c-4a51-b633-911ad37172c9.png) |

## To test
- Enable the feature flag is not active
- Go to Me screen and check the menu is there
- Set airplane mode ON and check that tapping the share menu returns a snackbar for no network
- Set airplane mode OFF and check that now the share sheet is shown and you can share the link with text (and subject if using the email to share)
- Repeat the above configuring only a self-hosted site in the app, everything should work the same.
- Smoke test the me screen

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
No impact expected but did some manual smoke testing regardless.

3. What automated tests I added (or what prevented me from doing so)
Added a unit tests to the MeViewModel.

PR submission checklist:
